kotlin/generate-offline-map/src/main/java/com/esri/arcgisruntime/sample/generateofflinemap/MainActivity.kt

```diff
@@ -0,0 +1,231 @@
+    var minScale: Double = mapView.mapScale
+    val maxScale: Double = mapView.map.maxScale
+    // minScale must always be larger than maxScale
+    if (minScale <= maxScale) {
+      minScale = maxScale + 1
+    }
+
+    val generateOfflineMapParameters = GenerateOfflineMapParameters(
+      downloadArea.geometry, minScale, maxScale
+    )
+    // create an offline map task with the map
+    val offlineMapTask = OfflineMapTask(mapView.map)
+
+    // create an offline map job with the download directory path and parameters and start the job
+    val offlineMapJob =
+      offlineMapTask.generateOfflineMap(generateOfflineMapParameters, tempDirectoryPath)
+
+    // create an alert dialog to show the download progress
+    val progressDialog = createProgressDialog(offlineMapJob)
+    progressDialog.show()
+
+    offlineMapJob.apply {
+      // link the progress bar to the job's progress
+      addProgressChangedListener {
+        progressDialog.progressBar.progress = progress
+        progressDialog.progressTextView.text = "${progress}%"
+      }
+
+      // replace the current map with the result offline map when the job finishes
+      addJobDoneListener {
+        if (status == Job.Status.SUCCEEDED) {
+          val result = result
+          mapView.map = result.offlineMap
+          graphicsOverlay.graphics.clear()
+
+          // disable and remove the button to take the map offline once the offline map is showing
+          takeMapOfflineButton.isEnabled = false
+          takeMapOfflineButton.visibility = View.GONE
+
+          Toast.makeText(this@MainActivity, "Now displaying offline map.", Toast.LENGTH_LONG).show()
+        } else {
+          val error =
+            "Error in generate offline map job: " + offlineMapJob.error.additionalMessage
+          Toast.makeText(this@MainActivity, error, Toast.LENGTH_LONG).show()
+          Log.e(TAG, error)
+        }
+        // close the progress dialog
+        progressDialog.dismiss()
+      }
+      // start the job
+      start()
+    }
+  }
+
+  /**
+   * Create a progress dialog box for tracking the generate offline map job.
+   *
+   * @param job the generate offline map job progress to be tracked
+   * @return an AlertDialog set with the dialog layout view
+   */
+  private fun createProgressDialog(job: GenerateOfflineMapJob): AlertDialog {
+    val builder = AlertDialog.Builder(this@MainActivity).apply {
+      setTitle("Generating offline map...")
+      // provide a cancel button on the dialog
+      setNeutralButton("Cancel") { _, _ ->
+        job.cancel()
+      }
+      setCancelable(false)
+      setView(
+        LayoutInflater.from(this@MainActivity)
+          .inflate(R.layout.dialog_layout, null)
+      )
+    }
+    return builder.create()
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    // delete app cache when the app loses focus
+    cacheDir.deleteRecursively()
```

Comment:
```suggestion
```

@TADraeseke I wrote a note to myself to remove this line because cache stuff will be handled by the sample viewer. However I'm not sure now if that applies because this cache is set up manually in the generateOfflineMap() function (which I brought over from the java)...any ideas?

Reply:
I think I've done that correctly now